# Airfare Analytics — Data Understanding

This notebook provides an initial exploration of the raw flight data, focusing on dataset structure, data quality, variable types, and potential issues before conducting the exploratory data analysis.

In [34]:
# Import the libraries necessaries for this project
import pandas as pd
import numpy as np

In [35]:
# Upload the datasets from the raw folder
airfare_df = pd.read_csv("../data/raw/europe_flights_google_prices.csv")

In [36]:
# Check the first 5 rows of the dataset
airfare_df.head()


,airline,flight,source_city,source_country,departure_time,stops,arrival_time,destination_city,destination_country,class,price,days_left,duration,source_airport,destination_airport,scraped_at,departure_daypart,arrival_daypart,departure_day_of_week,arrival_day_of_week
0,easyJet,U28122,César Manrique-Lanzarote Airport,Spain,2026-05-15T19:15,0,2026-05-15T23:15,London Gatwick Airport,United Kingdom,economy,19,14,4h 00m,ACE,LGW,2026-05-01T21:23:11+0300,evening,night,friday,friday
1,easyJet,U28126,César Manrique-Lanzarote Airport,Spain,2026-05-15T13:45,0,2026-05-15T17:50,London Gatwick Airport,United Kingdom,economy,32,14,4h 05m,ACE,LGW,2026-05-01T21:23:11+0300,afternoon,evening,friday,friday
2,British Airways,BA2705,César Manrique-Lanzarote Airport,Spain,2026-05-15T11:45,0,2026-05-15T15:35,London Gatwick Airport,United Kingdom,economy,119,14,3h 50m,ACE,LGW,2026-05-01T21:23:11+0300,morning,afternoon,friday,friday
3,Iberia,IB1502 / IB1863,César Manrique-Lanzarote Airport,Spain,2026-05-15T11:15,1,2026-05-15T19:55,London Gatwick Airport,United Kingdom,economy,134,14,4h 55m,ACE,LGW,2026-05-01T21:23:11+0300,morning,evening,friday,friday
4,Iberia,IB1864 / IB1501,London Gatwick Airport,United Kingdom,2026-05-15T20:40,1,2026-05-16T10:30,César Manrique-Lanzarote Airport,Spain,economy,149,14,5h 04m,LGW,ACE,2026-05-01T21:23:13+0300,evening,morning,friday,saturday


In [37]:
# Check the information of the dataset
airfare_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 200002 entries, 0 to 200001
Data columns (total 20 columns):
 #   Column                 Non-Null Count   Dtype
---  ------                 --------------   -----
 0   airline                200002 non-null  str  
 1   flight                 200002 non-null  str  
 2   source_city            200002 non-null  str  
 3   source_country         200002 non-null  str  
 4   departure_time         200002 non-null  str  
 5   stops                  200002 non-null  int64
 6   arrival_time           200002 non-null  str  
 7   destination_city       200002 non-null  str  
 8   destination_country    200002 non-null  str  
 9   class                  200002 non-null  str  
 10  price                  200002 non-null  int64
 11  days_left              200002 non-null  int64
 12  duration               200002 non-null  str  
 13  source_airport         200002 non-null  str  
 14  destination_airport    200002 non-null  str  
 15  scraped_at             20000

In [38]:
# Check for missing values and the number of unique values in each column
quality = pd.DataFrame({
    'missing_values': airfare_df.isna().sum(),
    'missing_pct': (airfare_df.isna().mean() * 100).round(2),
    'n_unique': airfare_df.nunique()
})
quality

,missing_values,missing_pct,n_unique
airline,0,0.0,1214
flight,0,0.0,71811
source_city,0,0.0,269
source_country,0,0.0,41
departure_time,0,0.0,14269
stops,0,0.0,6
arrival_time,0,0.0,14028
destination_city,0,0.0,263
destination_country,0,0.0,40
class,0,0.0,1


In [39]:
# Check the number of rows and columns in the dataset
print(f"Rows: {airfare_df.shape[0]}")
print(f"Columns: {airfare_df.shape[1]}")

airfare_df.columns.tolist()

Rows: 200002
Columns: 20


['airline',
 'flight',
 'source_city',
 'source_country',
 'departure_time',
 'stops',
 'arrival_time',
 'destination_city',
 'destination_country',
 'class',
 'price',
 'days_left',
 'duration',
 'source_airport',
 'destination_airport',
 'scraped_at',
 'departure_daypart',
 'arrival_daypart',
 'departure_day_of_week',
 'arrival_day_of_week']

In [40]:
# Check the number of unique values in each column
duplicate_rate = airfare_df.duplicated().mean() * 100
print(f'Duplicate rate: {duplicate_rate:.4f}%')

Duplicate rate: 0.0005%


In [41]:
airfare_df['source_country'].unique()


<StringArray>
[         'Spain', 'United Kingdom',        'Germany',         'France',
        'Albania',          'Italy',         'Sweden',        'Austria',
        'Denmark',    'Netherlands',    'Switzerland',        'Belgium',
         'Russia', 'Czech Republic',         'Greece',       'Portugal',
         'Turkey',        'Finland',        'Croatia',     'Montenegro',
         'Poland',         'Norway',         'Serbia',        'Hungary',
         'Latvia',       'Slovakia',       'Bulgaria',         'Jersey',
        'Romania',     'Luxembourg',        'Ireland',        'Iceland',
          'Malta',      'Lithuania',      'Macedonia',       'Slovenia',
        'Estonia',        'Belarus',      'Gibraltar',       'Guernsey',
    'Isle of Man']
Length: 41, dtype: str

In [42]:
# Check the unique values in the 'airline' column
airfare_df['airline'].unique()

<StringArray>
[                                                'easyJet',
                                         'British Airways',
                                                  'Iberia',
                                                 'Vueling',
                                              'Air Europa',
                                  'canaryfly + Air Europa',
                            'Binter Canarias + Air Europa',
                                                    'Jet2',
                       'Binter Canarias + British Airways',
                                               'Eurowings',
 ...
             'Brussels Airlines + SWISS + British Airways',
                   'Scandinavian Airlines + Qatar Airways',
                           'SWISS + Austrian + SunExpress',
 'Austrian + Air Dolomiti + Lufthansa + Discover Airlines',
                          'Air Baltic + KM Malta Airlines',
                                          'Emirates + ITA',
     'SWISS + Lufthan

In [43]:
# Check the distribution of the 'price' column
airfare_df["price"].describe()

count    200002.000000
mean        384.010825
std         395.301844
min          17.000000
25%         189.000000
50%         289.000000
75%         455.000000
max       11614.000000
Name: price, dtype: float64

In [44]:
# Check the range of the 'departure_time' column
departure_dt = pd.to_datetime(airfare_df["departure_time"])
print("First departure:", departure_dt.min())
print("Last departure:", departure_dt.max())

First departure: 2026-05-01 19:30:00
Last departure: 2026-06-30 23:55:00


## Conclusions

### Dataset Overview

- The dataset contains 200,002 flight observations and 20 variables.
- It covers flights departing between May 1, 2026 and June 30, 2026.
- The dataset includes information about airlines, routes, airports, departure and arrival times, number of stops, flight duration, booking lead time, and ticket prices.
- The target variable for the predictive modeling stage will be 'price'.

### Data Quality

- No missing values were detected across any of the 20 variables.
- The duplicate rate is approximately 0.0005%, indicating that duplicated observations are negligible.
- Several temporal variables (departure_time , arrival_time, and scraped_at) are currently stored as strings and should be converted to datetime format.
- 'duration' is also stored as a string and will require inspection and potentially conversion into a numerical representation.

### Dataset Characteristics

- The dataset contains flights from 41 source countries and 40 destination countries, providing broad geographical coverage across Europe.
- 'days_left' contains 61 distinct values and represents the booking lead time before departure.
- Departure and arrival dayparts contain four categories, while day-of-week variables contain the expected seven categories.

### Target Variable

- Ticket prices range from 17 to 11,614.
- The median ticket price is 289, while the mean is approximately 384.

